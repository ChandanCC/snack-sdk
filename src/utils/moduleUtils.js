/* @flow */

import trim from 'lodash/trim';
import { types, parse, print } from 'recast';
import * as babylon from 'babylon';
import config from '../configs/babylon';

const parser = {
  parse: (code: string) => babylon.parse(code, config),
};

const b = types.builders;

const getModuleNameFromRequire = (node: *) => {
  const { callee, arguments: args } = node;

  let name;

  if (
    callee.name === 'require' &&
    args.length === 1 &&
    (args[0].type === 'StringLiteral' || args[0].type === 'TemplateLiteral')
  ) {
    if (args[0].type === 'StringLiteral') {
      name = args[0].value;
    } else if (args[0].type === 'TemplateLiteral' && args[0].quasis.length === 1) {
      name = args[0].quasis[0].value.cooked;
    }
  }

  return name && !/\n/.test(name) && !name.startsWith('/') ? name : null;
};

const getVersionFromComments = (comments: *) => {
  // Makes sure version specified is valid
  const versionFormat = /^(\s*(\d+\.)?(\d+\.)?(\*|\d+))|(LATEST)$/;

  return comments && versionFormat.test(comments[0].value)
    ? trim(comments[0].value.replace(/\s*/, ''))
    : null;
};

const addCommentToPath = (path, comment: string) => {
  const source = print(path).code;
  const lineComment = `// ${comment}`;
  const blockComment = `/* ${comment} */`;

  if (source.endsWith(lineComment) || source.endsWith(blockComment)) {
    return;
  }

  path.replace(parse(`${source} ${lineComment}`, { parser }).program.body[0]);
};

export const findModuleDependencies = (code: string): { [string]: string } => {
  const ast = parse(code, { parser });
  const dependencies: { [string]: string } = {};

  types.visit(ast, {
    visitImportDeclaration(path) {
      dependencies[path.node.source.value] = getVersionFromComments(path.node.trailingComments);

      this.traverse(path);
    },

    visitExportNamedDeclaration(path) {
      if (path.node.source) {
        dependencies[path.node.source.value] = getVersionFromComments(path.node.trailingComments);
      }
      this.traverse(path);
    },

    visitExportAllDeclaration(path) {
      dependencies[path.node.source.value] = getVersionFromComments(path.node.trailingComments);
      this.traverse(path);
    },

    visitCallExpression(path) {
      const name = getModuleNameFromRequire(path.node);

      if (name) {
        const { arguments: args } = path.node;

        // Comment location changes if user uses semicolons or not
        const version = getVersionFromComments(
          args[0].trailingComments || path.parentPath.parentPath.node.trailingComments
        );

        dependencies[name] = version;
      }

      this.traverse(path);
    },
  });

  return dependencies;
};

// Writes version number in comments in code
export const writeModuleVersions = (code: string, dependencies: { [string]: string }): string => {
  const ast = parse(code, { parser });

  types.visit(ast, {
    visitImportDeclaration(path) {
      const name = path.node.source.value;
      const version = dependencies[name];

      if (version) {
        addCommentToPath(path, version);
      }

      return false;
    },

    visitExportNamedDeclaration(path) {
      const name = path.node.source && path.node.source.value;
      const version = dependencies[name];

      if (version) {
        addCommentToPath(path, version);
      }

      return false;
    },

    visitExportAllDeclaration(path) {
      const name = path.node.source.value;
      const version = dependencies[name];

      if (version) {
        addCommentToPath(path, version);
      }

      return false;
    },

    visitCallExpression(path) {
      const name = getModuleNameFromRequire(path.node);

      if (name) {
        const version = dependencies[name];

        if (version) {
          addCommentToPath(path, version);
        }
      }

      return false;
    },
  });

  return print(ast).code;
};

