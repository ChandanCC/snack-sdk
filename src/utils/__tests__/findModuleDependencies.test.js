/* eslint-env jest */

import moduleUtils from '../findModuleDependencies';

it('finds all imported modules', () => {
  const code = `
    import base64 from 'base64'; // 1.2.3
    import debounce from 'lodash/debounce'; // 2.3.4
    import { connect } from 'react-redux';
  `;

  const dependencies = moduleUtils.findModuleDependencies(code);

  expect(dependencies).toMatchSnapshot();
});

it('finds all required modules', () => {
  const code = `
    const base64 = require('base64'); // 1.2.3
    const debounce = require('lodash/debounce'); // 2.3.4
    const { connect } = require('react-redux');
  `;

  const dependencies = moduleUtils.findModuleDependencies(code);

  expect(dependencies).toMatchSnapshot();
});

it('writes versions for imported modules', () => {
  const code = `
    import base64 from 'base64';
    import debounce from 'lodash/debounce';
    import { connect } from 'react-redux';
  `;

  const modules = {
    base64: {
      name: 'base64',
      version: '1.2.3',
    },
    'lodash/debounce': {
      name: 'lodash/debounce',
      version: '2.3.4',
    },
  };

  const dependencies = moduleUtils.writeModuleVersions(code, modules);

  expect(dependencies).toMatchSnapshot();
});

it('writes versions for required modules', () => {
  const code = `
    const base64 = require('base64');
    const debounce = require('lodash/debounce');
    const { connect } = require('react-redux');
  `;

  const modules = {
    base64: {
      name: 'base64',
      version: '1.2.3',
    },
    'lodash/debounce': {
      name: 'lodash/debounce',
      version: '2.3.4',
    },
  };

  const dependencies = moduleUtils.writeModuleVersions(code, modules);

  expect(dependencies).toMatchSnapshot();
});
