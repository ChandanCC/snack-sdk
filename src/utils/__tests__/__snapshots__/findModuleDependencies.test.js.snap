exports[`test finds all imported modules 1`] = `
Array [
  Object {
    "name": "base64",
    "version": "1.2.3",
  },
  Object {
    "name": "lodash/debounce",
    "version": "2.3.4",
  },
  Object {
    "name": "react-redux",
    "version": null,
  },
]
`;

exports[`test finds all required modules 1`] = `
Array [
  Object {
    "name": "base64",
    "version": "1.2.3",
  },
  Object {
    "name": "lodash/debounce",
    "version": "2.3.4",
  },
  Object {
    "name": "react-redux",
    "version": null,
  },
]
`;

exports[`test writes versions for imported modules 1`] = `
"
    import base64 from \'base64\'; // 1.2.3
    import debounce from \'lodash/debounce\'; // 2.3.4
    import { connect } from \'react-redux\';
  "
`;

exports[`test writes versions for required modules 1`] = `
"
    const base64 = require(\'base64\'); // 1.2.3
    const debounce = require(\'lodash/debounce\'); // 2.3.4
    const { connect } = require(\'react-redux\');
  "
`;
